Replace biopython with faster dnaio, buffer writes to reduce memory spikes during renaming

bede opened the pull request:
This PR contains necessary modifications to read renaming in order to use detaxizer with large datasets (50-100GB compressed) on a machine with 128GB of RAM. Pipeline execution time is also dramatically reduced. Please do not feel obliged to merge, but I wanted to share my changes in case they help someone.

- Use buffered writes to prevent unbounded growth of the read renaming dictionary
- Use dnaio by @marcelm instead of Biopython for FASTQ parsing, which is >10x faster
- Change process label to `process_low` since memory usage is now bounded

Test profile runs successfully with conda and docker.

Should you consider merging, I will update the changelog etc. accordingly. These modifications appear to be working well for me, but it's possible renaming has regressed e.g. wrt blocked gzip support.

d4straub commented:
Thanks a lot for sharing your improvement! The speed of this step is a bottleneck, indeed.
Any improvement, especially easing this bottleneck, is definitely welcome.
Is blocked gzip not supported by dnaio? Or why do you think such a problem might occur?
If indeed dnaio doesnt support blocked gzip but the current solution does (sorry, have no idea), then it might be better to have both solutions available?

marcelm commented:
Since I’ve been tagged: dnaio supports reading concatenated/multiblock gzip files.

d4straub commented:
> Since I’ve been tagged: dnaio supports reading concatenated/multiblock gzip files.

Great, thanks, so adding this improvement doesn't need to be alongside the old implementation.
I am still a bit concerned about the part `but it's possible renaming has regressed`, I am not a python person unfortunately and cannot really judge.

Unfortunately all that failing tests are from the outdated nf-core template, see https://github.com/nf-core/detaxizer/pull/60

bede commented:
Thanks @d4straub, I mentioned the possibility of regression out of caution, not because I'm aware of issues. Used with custom config specifying memory limits, this PR was the first version that completed successfully for all of my test datasets in a single execution without retries or crashes. If there is appetite to merge, I could also look into applying prerequisite nf-core template updates, but likely not for another week or two.

d4straub commented:
It would be very nice to have that improvement in the pipeline. So yes, if you find the time it would be great to update template & merge the PR (to dev though, not master, at least as first step).

d4straub commented:
Dear @bede , I'll attempt the template update in the next days, just fyi so that we dont do overlapping work here.

d4straub commented:
Dear @bede , the pipeline recieved some changes and I would like to make a release soon. I think the pipeline would benefit from your additions. Do you want to work on that PR or may I use/copy your code and add it to the pipeline in a different PR (I can request you as reviewer if possible).

bede commented:
Hi @d4straub. You could add commits to this PR branch in order to apply the template changes, or else would you like me to do it? I have never made nf-core template updates so would have to learn, but I would rather this than this my work on this PR going to waste. Thanks

bede commented:
Hi Daniel,
I've rebased the changes onto the tip of dev, and added a citation and changelog entry following your example. Assuming this passes tests, are you happy to close #74 and merge this instead? I'm happy to make further minor changes, otherwise feel free to checkout this PR branch and add commits before merging. Thank you.

d4straub commented:
Dear Bede, all tests passed and you are good to merge. I prefer the PR opener to merge their own PRs. Potentially you need to join the nf-core github organisation via https://nf-co.re/join#github in order to merge this PR. If you want me to merge the PR instead, please let me know.

bede commented:
Thank you for patience Daniel, I have been on holidays since Friday and will try to merge this evening.
